File: year2019/c++/day4/include/day4_func.cpp
#include "day4_func.h"
#include "../../../../external/cppitertools/itertools.hpp"

PasswordFinder::PasswordFinder(int from, int to) : from(from), to(to) {}

std::vector<int> PasswordFinder::getPasswords() {
    std::vector<int> passwords;
    for(int i = from; i <= to; i++) {
        passwords.push_back(i);
    }
    return passwords;
};

bool IncreasingValidation::Validate(std::string password) {
    bool less = false;
    for(int i = 0; i < password.size() - 1; i++) {
        less = password[i] <= password[i+1];
        if (less == false) {
            break;
        }
    }
    return less;
};

bool TwoSequentuallyEqual::Validate(std::string password) {
    bool equal = false;
    for(int i = 0; i < password.size() - 1; i++) {
        equal = password[i] == password[i+1];
        if (equal) {
            break;
        }
    }
    return equal;
};

bool TwoEqual::Validate(std::string password) {
    bool twoEqual = false;

    for (auto&& gb : iter::groupby(password)) {
        int count = 0;
        for (auto&& s : gb.second) {
            count++;
        }

        if(count == 2) {
            twoEqual = true;
            break;
        }
    }

    return twoEqual;    
}

PasswordValidator::~PasswordValidator()
{
    this->validations.clear();
}

void PasswordValidator::setPasswordValidations(std::vector<Validation*> validations) {
    this->validations = validations;
}

bool PasswordValidator::validatePassword(int password){
    std::string passwordString = std::to_string(password);
    bool validPassword = false;
    for(int i = 0; i < validations.size(); i++) {
        validPassword = validations[i]->Validate(passwordString);
        if(!validPassword) {
            break;
        }
    }
    return validPassword;
};

int PasswordValidator::numberValidPassword(std::vector<int> passwords) {
    int numberValid = 0;
    std::for_each(std::begin(passwords), std::end(passwords),
     [this, &numberValid](int password){
         numberValid += this->validatePassword(password) ? 1 : 0;
     });
     return numberValid;
}

File: year2019/python/day4/day4_func.py
from __future__ import annotations
from itertools import groupby
from abc import ABC, abstractmethod
from typing import List


class PasswordFinder:
    def __init__(self, start: int, end: int) -> None:
        self.start = start
        self.end = end

    def getPasswords(self):
        return [s for s in range(self.start, self.end + 1)]


class Validation(ABC):

    @abstractmethod
    def validate(self, password: str) -> bool:
        pass


class IncreasingValidation(Validation):

    def validate(self, password: str) -> bool:
        return all([a <= b for a, b in zip(password, password[1:])])


class TwoSequentuallyEqual(Validation):

    def validate(self, password: str) -> bool:
        return any([a == b for a, b in zip(password, password[1:])])


class TwoEqual(Validation):

    def validate(self, password: str) -> bool:
        return any([len(list(k)) == 2 for _, k in groupby(password)])


class PasswordValidator:

    def __init__(self):
        self._validations = None

    @property
    def validations(self):
        return self._validations

    @validations.setter
    def validations(self, validationsInput: List[Validation]) -> None:
        self._validations = validationsInput

    def validatePassword(self, password: int):
        return all(validation.validate(str(password)) for validation in self.validations)

    def numberValidPassword(self, passwords: List[int]):
        return sum([self.validatePassword(password) for password in passwords])
